### PR TITLE
[2.x] Backport query assist

### PR DESCRIPTION
**Pull request by joshuali925**
### Description
backports #1325, #1359, #1360, #1376, #1379

cc @paulstn i picked the 3 commits for query assist to 2.x, if you have other fixes for it please let me know or update this PR

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/plugin.ts [hunk 1]
```diff
@@ -64,6 +65,9 @@ import {
   OBSERVABILITY_EMBEDDABLE_ID,
 } from './embeddable/observability_embeddable';
 import { ObservabilityEmbeddableFactoryDefinition } from './embeddable/observability_embeddable_factory';
+import { coreRefs } from './framework/core_refs';
+import { S3DataSource } from './framework/datasources/s3_datasource';
```
Comment (ps48) [inline public/plugin.ts:69]: Looks like a double import is causing CI failure.

Comment (joshuali925) [reply public/plugin.ts:68]: thanks, didn't get time to test yesterday

**Comment by RyanL1997** [comment]
Thanks for taking this on @joshuali925. LGTM, and just a tiny linting error need to be fixed: https://github.com/opensearch-project/dashboards-observability/actions/runs/7588601129/job/20671492177?pr=1369#step:10:95

**Review by RyanL1997 — APPROVED** [review]

**Comment by joshuali925** [comment]
@RyanL1997 i just realized that current lint workflow will fail if one file failed to lint, it gets tedious if there are multiple failed files since it requires multiple runs

i sent a PR to fix it https://github.com/opensearch-project/dashboards-observability/pull/1378

**Review by Swiddis — APPROVED** [review]

**Review by paulstn — APPROVED** [review]

**Pull request merged**